engine/src/main/java/org/hibernate/search/spi/impl/HashSetIndexedTypesSet.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.spi.impl;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.hibernate.search.spi.IndexedTypeIdentifier;
+import org.hibernate.search.spi.IndexedTypesSet;
+
+final class HashSetIndexedTypesSet implements IndexedTypesSet, Serializable {
+
+	public static final IndexedTypesSet EMPTY = new HashSetIndexedTypesSet( Collections.EMPTY_SET );
+
+	private final Set<IndexedTypeIdentifier> set;
+
+	HashSetIndexedTypesSet(Set<IndexedTypeIdentifier> set) {
```

Comment:
If you don't want to change the implementation (see my comment above), this should only accept `HashSet<IndexedTypeIdentifier>`.